File: examples/runtimes/ruby/lib/nodes/hello_world.rb
# frozen_string_literal: true

require_relative "../node_handler"

module Blok
  module Nodes
    # HelloWorldNode is an example Blok node in Ruby.
    #
    # It reads a +name+ from the request body and an optional +prefix+
    # from the node config, then returns a greeting message.
    class HelloWorldNode < Blok::NodeHandler
      # Execute the hello-world node.
      #
      # @param ctx [Blok::Context] the workflow execution context
      # @param config [Hash] the node configuration (supports "prefix" key)
      # @return [Hash] a hash containing the greeting message, timestamp, and language
      def execute(ctx, config)
        # Get name from request body or use default
        name = "World"
        body = ctx.request.body
        if body.is_a?(Hash) && body.key?("name")
          name = body["name"].to_s
        end

        # Get greeting prefix from config or use default
        prefix = config&.fetch("prefix", nil) || "Hello"

        message = "#{prefix}, #{name}!"

        # Store in context vars for downstream nodes
        ctx.vars["greeting"]  = message
        ctx.vars["timestamp"] = Time.now.to_i

        # Return response
        {
          "message"   => message,
          "timestamp" => Time.now.iso8601,
          "language"  => "Ruby"
        }
      end
    end
  end
end

File: examples/runtimes/rust/src/nodes/hello_world.rs
use async_trait::async_trait;
use std::collections::HashMap;

use crate::node::NodeHandler;
use crate::types::Context;

/// HelloWorldNode is an example Blok node implemented in Rust.
///
/// It reads an optional `name` from the request body, an optional `prefix`
/// from the node config, and returns a greeting message.
pub struct HelloWorldNode;

#[async_trait]
impl NodeHandler for HelloWorldNode {
    async fn execute(
        &self,
        ctx: &mut Context,
        config: &HashMap<String, serde_json::Value>,
    ) -> Result<serde_json::Value, Box<dyn std::error::Error + Send + Sync>> {
        // Get name from request body or use default
        let name = ctx
            .request
            .body
            .get("name")
            .and_then(|v| v.as_str())
            .unwrap_or("World");

        // Get greeting prefix from config or use default
        let prefix = config
            .get("prefix")
            .and_then(|v| v.as_str())
            .unwrap_or("Hello");

        let message = format!("{}, {}!", prefix, name);

        // Store in context vars for downstream nodes
        ctx.vars.insert(
            "greeting".to_string(),
            serde_json::Value::String(message.clone()),
        );
        ctx.vars.insert(
            "timestamp".to_string(),
            serde_json::json!(chrono::Utc::now().timestamp()),
        );

        // Return response
        Ok(serde_json::json!({
            "message": message,
            "timestamp": chrono::Utc::now().to_rfc3339(),
            "language": "Rust"
        }))
    }
}
